### ISSUE TITLE
Issues with scanning Avalon miners

### ISSUE THREAD

**Issue by Nelly1234567** [UpstreamData/pyasic#59]
I've been looking for ways to make reading the Canaan Avalon logs easier and found this python library. However, there seems to be an issues with this scanning Avalon 1246 miners.

When scanning a single or range of Avalon 1246 miners, I get the following error:

```
Traceback (most recent call last):
  File "C:\Users\WM\Desktop\CanaanLogger\Temppyasic.py", line 16, in <module>
    asyncio.run(gather_miner_data())
  File "C:\Users\WM\AppData\Local\Programs\Python\Python311\Lib\asyncio\runners.py", line 190, in run
    return runner.run(main)
           ^^^^^^^^^^^^^^^^
  File "C:\Users\WM\AppData\Local\Programs\Python\Python311\Lib\asyncio\runners.py", line 118, in run
    return self._loop.run_until_complete(task)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "C:\Users\WM\AppData\Local\Programs\Python\Python311\Lib\asyncio\base_events.py", line 653, in run_until_complete
    return future.result()
           ^^^^^^^^^^^^^^^
  File "C:\Users\WM\Desktop\CanaanLogger\Temppyasic.py", line 10, in gather_miner_data
    all_miner_data = await asyncio.gather(*[miner.get_data() for miner in miners])
                     ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "C:\Users\WM\Desktop\CanaanLogger\venv\Lib\site-packages\pyasic\miners\base.py", line 536, in get_data
    gathered_data = await self._get_data(allow_warning, data_to_get=data_to_get)
                    ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "C:\Users\WM\Desktop\CanaanLogger\venv\Lib\site-packages\pyasic\miners\base.py", line 489, in _get_data
    miner_data[data_name] = await function(**args_to_send)
                            ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "C:\Users\WM\Desktop\CanaanLogger\venv\Lib\site-packages\pyasic\miners\backends\cgminer_avalon.py", line 186, in get_hostname
    return f"Avalon{mac.replace(':', '')[-6:]}"
                    ^^^^^^^^^^^
AttributeError: 'dict' object has no attribute 'replace'
```

I've been testing it on a small range of miners available to me and it appears to work fine for Bitmain S19 Pros, S19J Pros and T19's, just not for Avalon miners currently. The code I've been using is the example from the pyasic read the docs site for collecting all the data from a miner. 

```
import asyncio  # asyncio for handling the async part
from pyasic.network import MinerNetwork  # miner network handles the scanning


async def gather_miner_data():  # define async scan function to allow awaiting
    network = MinerNetwork("192.168.1.50")
    miners = await network.scan_network_for_miners()

    # we need to asyncio.gather() all the miners get_data() functions to make them run together
    all_miner_data = await asyncio.gather(*[miner.get_data() for miner in miners])

    for miner_data in all_miner_data:
        print(miner_data)    # print out all the data one by one

if __name__ == "__main__":
    asyncio.run(gather_miner_data())
```

I'm on Windows 10 using Python 3.11.3. I tested the example GUI program shown on the "https://pypi.org/project/pyasic/" page which appears to work for the most part with the Avalon 1246 miners but it does seem to be using a much earlier version of pyasic. (v0.35.0 I believe).

As stated, the Miner's having issues in question is the Avalon 1246 miners (Avalon N 1246 as well which is pretty much the same thing)
Firmware version is "21042001_4ec6bb0_61407fa" and "22122701_66620f1_fcec056"

**Comment by UpstreamData** [comment]
Looking at the trace back, the issue is with the `get_hostname` function being called by `get_data`.  It should likely just be removed, as that method isn't technically correct for some models, but it seems like `get_Mac` is returning a dict when it should be returning a str, so it seems like the root issue may be there.

Can you send the results of `await miner.get_mac()` and `await miner.api.send_command("version", ignore_errors=True)` so I can try to troubleshoot `get_mac`?

**Comment by Nelly1234567** [comment]
This is what I get using those:

```
B4:A2:EB:35:34:0B
{'STATUS': [{'STATUS': 'S', 'When': 540240, 'Code': 22, 'Msg': 'CGMiner versions', 'Description': 'cgminer 4.11.1'}], 'VERSION': [{'CGMiner': '4.11.1', 'API': '3.7', 'STM8': '20.08.01', 'PROD': 'AvalonMiner 1246-85', 'MODEL': '1246-85', 'HWTYPE': 'MM3v2_X3', 'SWTYPE': 'MM314', 'VERSION': '21042001_4ec6bb0_61407fa', 'LOADER': 'd0d779de.00', 'DNA': '0201000076f74997', 'MAC': 'b4a2eb35340b', 'UPAPI': '2'}], 'id': 1}
```

**Comment by Nelly1234567** [comment]
Here's one from a Avalon N 1246 as well if it helps any:

```
B4:A2:EB:3D:7A:3C

{'STATUS': [{'STATUS': 'S', 'When': 540495, 'Code': 22, 'Msg': 'CGMiner versions', 'Description': 'cgminer 4.11.1'}], 'VERSION': [{'CGMiner': '4.11.1', 'API': '3.7', 'STM8': '20.08.01', 'PROD': 'AvalonMiner 1246-N-90', 'MODEL': '1246-N-90', 'HWTYPE': 'MM3v2_X3', 'SWTYPE': 'MM315', 'VERSION': '22101101_66620f1_f39b73b', 'LOADER': 'd0d779de.00', 'DNA': '0201000056ae1375', 'MAC': 'b4a2eb3d7a3c', 'UPAPI': '2'}], 'id': 1}
```

**Comment by UpstreamData** [comment]
Looks correct.  I'll pull hostname out.

You can use it like this instead for now - 
```
data_to_get = list(miner.data_locations.keys())
data_to_get.remove("hostname")
await miner.get_data(data_to_get=data_to_get)
```

**Comment by Nelly1234567** [comment]
Perfect, that works for me so far. Thanks for the help!

**Comment by UpstreamData** [comment]
Should be fixed in `v0.38.0`.  Also, the name of the attribute to select which data to gather has been changed to `include` and you can now use `exclude`, so the code would change to `await miner.get_data(exclude=["hostname"])`
